src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -87,6 +87,28 @@ impl PermissionPriv {
     fn compatible_with_protector(&self) -> bool {
         !matches!(self, ReservedIM)
     }
+
+    /// Returns the strongest foreign action this node survives (without change),
+    /// where `prot` indicates if it is protected.
+    /// They are ordered as `None` < `Some(Read)` < `Some(Write)`, in order of power.
+    pub fn strongest_survivable_foreign_action(&self, prot: bool) -> Option<AccessKind> {
+        match self {
+            // The read will make it conflicted, so it is not invariant under it.
+            ReservedFrz { conflicted } if prot && !conflicted => None,
+            // Otherwise, foreign reads do not affect it
+            ReservedFrz { .. } => Some(AccessKind::Read),
+            // Famously, ReservedIM survives foreign writes. It is never protected.
```

Comment:
Might we worth a (debug) assertion of `!prot`.

Reply:
Done